### ISSUE TITLE
NotImplementedError: lzw_encode

### ISSUE THREAD

**Issue by davideboschetto** [cgohlke/tifffile#34]
Hi there,
I'm trying to write a multi page tiff file, like this

`    with tifffile.TiffWriter("nyname.tif", append=True) as P:
        P.write(**this_page_dict)`

where the dict is:
```
{
'extratags': [(274, 3, 1, 1), (339, 3, 1, 1), (285, 2, 25, b'myTitle')], 
'bitspersample': 16,
'compression': 5, 
'photometric': 1, 
'rowsperstrip': 8, 
'planarconfig': 1, 
'resolution': (1677721600, 16777216), 
'data': array(..., dtype=uint16), 
'dtype': dtype('uint16')
}
```
It's giving me a NotImplemented error, since I'm using compression=5 (LZW). 

> Traceback (most recent call last):
  File "...filename.py", line 122, in <module>
    P.write(**this_page_dict)
  File "...\venv\lib\site-packages\tifffile\tifffile.py", line 2342, in write
    strip = compress(strip)
  File "imagecodecs\_imcd.pyx", line 856, in imagecodecs._imcd.lzw_encode
NotImplementedError: lzw_encode

Any pointer as of why and maybe how to "solve" / work around this?

**Comment by cgohlke** [comment]
The imagecodecs library does not currently support LZW encoding.

https://github.com/cgohlke/imagecodecs/blob/f38d0d4e5db1d08d92742a5d899325fc7c7b56e7/imagecodecs/_imcd.pyx#L854-L856

You could monkey patch imagecodecs with your own implementation of a `lzw_encode` function that takes a numpy array and returns bytes.

**Comment by noobOriented** [comment]
I try implement the encoding by reverse engineering `imagecodes.lzw_decode`:

```python
def lzw_encode(seq: bytes) -> bytes:
    CLEAR_TOKEN = 256
    END_TOKEN = 257
    # init table
    table = {bytes([i]): i for i in range(256)}
    table[object()] = CLEAR_TOKEN
    table[object()] = END_TOKEN
    assert len(table) == 258

    def gen_code():
        yield CLEAR_TOKEN
        s = b""
        for c in seq:
            c = bytes([c])
            sc = s + c
            if sc in table:
                s = sc
            else:
                yield table[s]
                table[sc] = len(table)
                s = c

        if s:
            yield table[s]
        yield END_TOKEN

    s = "".join(
        f"{code:0{len(table).bit_length()}b}"
        for code in gen_code()
    )
    x = int(s, 2)
    m = x.bit_length() % 8
    if m != 0:
        x <<= 8 - m

    return x.to_bytes(x.bit_length() // 8, 'big')
```

it passes the following test case

```python
def test_lzw_encode():
    data = np.random.randint(0, 255, size=1000, dtype=np.uint8).tobytes()

    encoded = lzw_encode(data)

    assert imagecodecs.lzw_decode(encoded) == data

```

feel free to use it, or maybe add it to imagecodecs?

**Comment by cookmscott** [comment]
LZW may be included in imagecodecs now: https://github.com/cgohlke/imagecodecs/blob/v2021.11.20/imagecodecs/_imcd.pyx#L1085-L1125

However, I've had no luck using it. Any direction here? Thank you!

**Comment by cgohlke** [comment]
The LZW encoder is still not implemented in imagecodecs. The placeholder encoder function raises a RuntimeError: `ImcdError: imcd_lzw_encode returned IMCD_LZW_NOTIMPLEMENTED`.

**Comment by cookmscott** [comment]
Hi @cgohlke , would you suggest using @noobOriented's work around? If so, I'm not quite sure how to implement that for LZW compression if you could include some notes 🙏

Also for sake of clarity, if imagecodecs is not ready to support LZW encoding, could we update the docs [here](https://github.com/cgohlke/tifffile/blame/ae84934b0e8a9f0acbd2cadd30b465d12c1952a3/README.rst#L51) and [here](https://github.com/cgohlke/tifffile/blame/ae84934b0e8a9f0acbd2cadd30b465d12c1952a3/README.rst#L205) to mention you can only decode with LZW via imagecodecs? Maybe imagecodecs will inlcude LZW soon (seems to be the case) and that's why this request is still open. Thank you!

**Comment by cgohlke** [comment]
Fixed in imagecodecs 2022.7.27.

**Comment by joanlyq** [comment]
@cgohlke Hi Chris, glad to know it's fixed in imagecodecs, but I'm still having the same error after I downloaded the new tifffile in a new conda environment. 
This is the command I used and I'm not too sure if I'm using the correct command flags. 

`tifffile.imwrite('output.tif', new_im, compression='lzw') `

here is my conda env 
`pillow                    9.2.0            py39h579eac4_0    conda-forge
tifffile                  2022.7.28          pyhd8ed1ab_0    conda-forge`

appreciate any help here.
